Sources/Swift/Tools/SentryLogBatcher.swift

```diff
@@ -6,24 +6,120 @@ import Foundation
 @_spi(Private) public class SentryLogBatcher: NSObject {
     
     private let client: SentryClient
+    private let flushTimeout: TimeInterval
+    private let maxBufferSizeBytes: Int
+    private let dispatchQueue: SentryDispatchQueueWrapper
     
-    @_spi(Private) public init(client: SentryClient) {
+    // All mutable state is accessed from the same dispatch queue.
+    private var encodedLogs: [Data] = []
+    private var encodedLogsSize: Int = 0
+    private var timerWorkItem: DispatchWorkItem?
+
+    @_spi(Private) public init(
+        client: SentryClient,
+        flushTimeout: TimeInterval,
+        maxBufferSizeBytes: Int,
+        dispatchQueue: SentryDispatchQueueWrapper
+    ) {
         self.client = client
+        self.flushTimeout = flushTimeout
+        self.maxBufferSizeBytes = maxBufferSizeBytes
+        self.dispatchQueue = dispatchQueue
         super.init()
     }
     
+    @_spi(Private) public convenience init(client: SentryClient, dispatchQueue: SentryDispatchQueueWrapper) {
+        self.init(
+            client: client,
+            flushTimeout: 5,
+            maxBufferSizeBytes: 1_024 * 1_024, // 1MB
+            dispatchQueue: dispatchQueue
+        )
+    }
+    
     func add(_ log: SentryLog) {
-        dispatch(logs: [log])
+        dispatchQueue.dispatchAsync { [weak self] in
+            self?.encodeAndBuffer(log: log)
+        }
     }
     
-    private func dispatch(logs: [SentryLog]) {
+    @objc
+    public func flush() {
+        dispatchQueue.dispatchAsync { [weak self] in
+            self?.performFlush()
+        }
+    }
+
+    // Helper
+
+    // Only ever call this from the dispatch queue.
+    private func encodeAndBuffer(log: SentryLog) {
         do {
-            let payload = ["items": logs]
-            let data = try encodeToJSONData(data: payload)
+            let encodedLog = try encodeToJSONData(data: log)
+            
+            let wasEmpty = encodedLogs.isEmpty
+            
+            encodedLogs.append(encodedLog)
+            encodedLogsSize += encodedLog.count
+            
+            let shouldFlushImmediatley = encodedLogsSize >= maxBufferSizeBytes
+            let shouldStartTimer = wasEmpty && timerWorkItem == nil && !shouldFlushImmediatley
             
-            client.captureLogsData(data)
+            if shouldStartTimer {
+                startTimer()
+            }
+            if shouldFlushImmediatley {
+                performFlush()
+            }
         } catch {
-            SentrySDKLog.error("Failed to create logs envelope.")
+            SentrySDKLog.error("Failed to encode log: \(error)")
+        }
+    }
+    
+    // Only ever call this from the dispatch queue.
+    private func startTimer() {
+        let timerWorkItem = DispatchWorkItem { [weak self] in
+            self?.performFlush()
```

Comment:
`l`: I think a debug log message that we're calling flush from the timer could be helpful.